libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,192 @@
+
+using namespace solidity::frontend;
+using namespace solidity::langutil;
+using namespace std::string_literals;
+using namespace std;
+
+namespace solidity::lsp
+{
+
+namespace
+{
+
+vector<Declaration const*> allAnnotatedDeclarations(Identifier const* _identifier)
+{
+	vector<Declaration const*> output;
+	output.push_back(_identifier->annotation().referencedDeclaration);
+	output += _identifier->annotation().candidateDeclarations;
+	return output;
+}
+
+}
+
+ReferenceCollector::ReferenceCollector(
+	Declaration const& _declaration,
+	std::string const& _sourceIdentifierName
+):
+	m_declaration{_declaration},
+	m_sourceIdentifierName{_sourceIdentifierName.empty() ? _declaration.name() : _sourceIdentifierName}
+{
+}
+
+std::vector<Reference> ReferenceCollector::collect(
+	Declaration const* _declaration,
+	ASTNode const& _ast,
+	std::string const& _sourceIdentifierName
+)
+{
+	if (!_declaration)
+		return {};
+
+	ReferenceCollector collector(*_declaration, _sourceIdentifierName);
+	_ast.accept(collector);
+	return std::move(collector.m_resultingReferences);
+}
+
+std::vector<Reference> ReferenceCollector::collect(
+	ASTNode const* _sourceNode,
+	SourceUnit const& _sourceUnit
+)
+{
+	if (!_sourceNode)
+		return {};
+
+	auto output = vector<Reference>{};
+
+	if (auto const* identifier = dynamic_cast<Identifier const*>(_sourceNode))
+	{
+		for (auto const* declaration: allAnnotatedDeclarations(identifier))
+			output += collect(declaration, _sourceUnit, identifier->name());
+	}
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(_sourceNode))
+	{
+		solAssert(identifierPath->path().size() >= 1, "");
+		output += collect(identifierPath->annotation().referencedDeclaration, _sourceUnit, identifierPath->path().back());
+	}
+	else if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(_sourceNode))
+		output += collect(memberAccess->annotation().referencedDeclaration, _sourceUnit, memberAccess->memberName());
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(_sourceNode))
+		output += collect(declaration, _sourceUnit, declaration->name());
+	else
+		lspAssert(false, ErrorCode::InternalError, "Unhandled AST node "s + typeid(*_sourceNode).name());
+
+	return output;
+}
+
+bool ReferenceCollector::visit(ImportDirective const& _import)
+{
+	if (_import.name() == m_sourceIdentifierName)
+		return true;
+	return false;
+}
+
+void ReferenceCollector::endVisit(ImportDirective const& _import)
+{
+	for (auto const& symbolAlias: _import.symbolAliases())
+	{
+		if (
+			m_sourceIdentifierName == *symbolAlias.alias &&
+			symbolAlias.symbol &&
+			symbolAlias.symbol->annotation().referencedDeclaration == &m_declaration
+		)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Read);
+			break;
+		}
+		else if (m_sourceIdentifierName == *symbolAlias.alias)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Text);
```

Comment:
I actually copy'n'pasted the documentation of the enum values into the .h file already. That's the comments on the right hand side. For ::Text it's saying "a textual occurence". Now, when having a look at the others, which are really clearly defined (read access or write access), the Text one to me feels like a thing to use if the other two don't match. like a string-search match to some text sequence which can find even stuff in comments. On the client side, most likely this will differenciate in color of the highlight. In my case writes are red, reads a green and textual is blue.

Thinking about it, that occurence should actually be Read and not Text, obviously.

Reply:
I think some of this explanation should be in these docstrings. The ones from the spec are really terse an unhelpful.

Reply:
@cameel I've been reading through these 3 messages again and again and - and now I'm a little bit confused of what else you want me to do. Before i'm touching this. let's have a voice call so we don't have 25 more forth and back in textual communication for a docstring. :)

Reply:
Sure, we can have a call.

But I simply meant that instead of (or maybe in addition to) just `a textual occurence`, our docstring should really say what is our interpretation of it and what we use this value for. Same for other values.